src/main/java/com/limechain/trie/structure/nibble/NibblesCollector.java

```diff
@@ -0,0 +1,46 @@
+package com.limechain.trie.structure.nibble;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+
+public class NibblesCollector implements Collector<Nibble, List<Nibble>, Nibbles> {
+    public static NibblesCollector toNibbles() {
+        return new NibblesCollector();
+    }
+
+    @Override
+    public Supplier<List<Nibble>> supplier() {
+        return LinkedList::new;
+    }
+
+    @Override
+    public BiConsumer<List<Nibble>, Nibble> accumulator() {
+        return List::add;
+    }
+
+    @Override
+    public BinaryOperator<List<Nibble>> combiner() {
+        return (list1, list2) -> {
+            list1.addAll(list2);
+            return list1;
+        };
+    }
+
+    @Override
+    public Function<List<Nibble>, Nibbles> finisher() {
+        // TODO: Avoid this unnecessary copying by introducing "lightweight by-reference only" constructors to Nibbles
+        return Nibbles::new;
```

Comment:
Again, on the topic of mutability. The current state for now is: everytime we construct a nibble, it clones the data structure passed in order to assure ownership of the nibbles. This might cause inefficiencies in places where copying isn't necessary, like when we're constructing the outer collection of nibbles for the sole purpose of wrapping it within a `Nibbles` instance... remains to be decided after we decide what we want `Nibbles` to be.

Reply:
This is also going in the above mentioned sync.